Find and replace

michael opened the pull request.

Commit:
WIP

Commit:
Fixes.

Commit:
Connect user interface with FindAndReplaceManager.

Commit:
Prototype tests for FindAndReplaceManager.

Commit:
Improve tests for find+replace.

Integral committed:
Find matches and collect selections.

Integral committed:
Replace all command.

Integral committed:
Replace all matches, separate matches computation.

Integral committed:
Test fixes.

Integral committed:
Replace next, improve matches computation.

Commit:
Find as you type.

Commit:
Improve find and replace.

obuchtala committed:
Refactored MarkersManager

obuchtala committed:
Update SpellChecker

obuchtala committed:
Render markers for find-and-replace results on-the-fly.

obuchtala committed:
Prevent infite loop in FindAndReplaceManager

obuchtala committed:
Allow for ready-made markers.

obuchtala committed:
FindAndReplace: Create markers from the beginning

obuchtala committed:
Silencio

Commit:
Fix regression in MarkersManager.

Commit:
Add ability to skip next selection rerender.

Commit:
Improve find and replace implementation.

Commit:
Update markers on document change.

obuchtala commented on packages/find-and-replace/FindAndReplaceManager.js:
```diff
@@ -209,20 +219,24 @@ class FindAndReplaceManager {
+    this._updateMarkers()
+    this.editorSession._setDirty('commandStates')
+    if (!renderSelection) {
+      this.editorSession.skipNextSelectionRerender()
```
I would prefer to implement this in the way I explained yesterday:
- introduce a 'focused' or 'blurred' state
- skip selection re-rendering if blurred

michael replied:
Ok. I was just concerned that you may end up trapped if some implementer forgets to unblur right after the performed action.

obuchtala replied:
Just don't think skipping next is enough. E.g. think if a reflow happens because of collaboration stuff. Then you would loose the cursor too.
I think the only valid implementation is to have an explicit app state. Even if that means that the app state needs to be managed carefully.

Commit:
Use explicit blurred state.

Commit:
Add proper tabindex values for find+replace panel.

Commit:
Use findAndReplaceManager directly.

Commit:
Make sure selection is rendered when FindPreviousCommand is triggered.

Commit:
Formatting.

Commit:
Closing of find and replace and clear markers.

obuchtala committed:
Export all package classes.

obuchtala committed:
Don't require tested classes directly.

ATTENTION: Tests now work on the real substance bundle,
so that tests better reflect real usage. For that you must
expose all necessary classes.

obuchtala committed:
Updated substance-bundler

Fixes #1042

Integral committed:
Test find and replace package importing.

Commit:
Always scroll to selected match.

Integral committed:
More test cases for find and replace.

Commit:
Add CloseFindAndReplaceCommand.

Integral committed:
Set selections for find next and find previous.

Integral committed:
Fix replacing.

Integral committed:
Fix and improve find and replace tests.

Integral committed:
Tool styling.

Integral committed:
Improve UI of find and replace panel.

Integral committed:
Don't set selection on start.

Integral committed:
Set closest to cursor match, fix replace last match error.

Commit:
Explicit blur state while search input fields have focus.

Commit:
Enable global key handlers.

Integral committed:
Set selections properly.

Integral committed:
Merge remote-tracking branch 'origin/find-and-replace' into find-and-replace

# Conflicts:
#	packages/find-and-replace/FindAndReplaceManager.js

Commit:
scrollToSelectedMatch on post render.

Commit:
Make _scrollToSelectedMatch less agressive.

This solves an issue where we would scroll back to the selected match
after we changed the selection.

Integral committed:
Replace all one by one from middle test.

Integral committed:
Fix replace from the middle.

Integral committed:
Fix negative match selection, don't compute closest match for null selection.

Commit:
Implement select() method for DOMElements.

Commit:
Disable surface tools in blurred state.

Commit:
Check if key has been handled, if yes stop event propagation.

Commit:
Improved find+replace.

Integral committed:
Debounce find input.

Integral committed:
Fix annotation command test.

Integral committed:
Check for a browser for DOM operations.

Integral committed:
Fixing lint errors.

Commit:
Use 'application/xml' when parsing XML

Commit:
Run lint and karma with 'npm test'

Integral commented:
Yay! Let's merge at last!

michael commented:
as long as it is green ;)

Integral commented:
Hurrah!
